File: app/src/main/java/com/leaguechampions/injection/AppModule.java
package com.leaguechampions.injection;

import android.content.Context;
import android.content.SharedPreferences;
import android.preference.PreferenceManager;

import com.google.gson.FieldNamingPolicy;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.leaguechampions.LeagueChampions;
import com.leaguechampions.data.remote.Api;
import com.leaguechampions.data.repository.ChampionRepository;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;

@Module //(includes = ViewModelModule.class)
public class AppModule {

    @Provides
    @Singleton
    Context providesContext(LeagueChampions application) {
        return application;
    }

    @Provides
    @Singleton
    SharedPreferences providesSharedPreferences(Context context) {
        return PreferenceManager.getDefaultSharedPreferences(context);
    }

    @Provides
    @Singleton
    Gson provideGson() {
        GsonBuilder gsonBuilder = new GsonBuilder();
        gsonBuilder.setFieldNamingPolicy(FieldNamingPolicy.LOWER_CASE_WITH_UNDERSCORES);
        return gsonBuilder.create();
    }

    // TODO put this somewhere else ? in DataModule ?
    @Provides
    @Singleton
    ChampionRepository provideChampionRepository(Api api) {
        return new ChampionRepository(api);
    }
}

File: app/src/main/java/com/leaguechampions/injection/AppModule.kt
package com.leaguechampions.injection

import android.content.Context
import android.content.SharedPreferences
import android.preference.PreferenceManager
import com.google.gson.FieldNamingPolicy
import com.google.gson.Gson
import com.google.gson.GsonBuilder
import com.leaguechampions.LeagueChampions
import com.leaguechampions.data.local.AppDatabase
import com.leaguechampions.features.champions.data.local.ChampionDao
import dagger.Module
import dagger.Provides
import javax.inject.Singleton

@Module //(includes = ViewModelModule.class)
class AppModule {

    @Provides
    @Singleton
    internal fun providesContext(application: LeagueChampions): Context {
        return application
    }

    @Provides
    @Singleton
    internal fun providesSharedPreferences(context: Context): SharedPreferences {
        return PreferenceManager.getDefaultSharedPreferences(context)
    }

    @Provides
    @Singleton
    internal fun provideGson(): Gson {
        val gsonBuilder = GsonBuilder()
        gsonBuilder.setFieldNamingPolicy(FieldNamingPolicy.LOWER_CASE_WITH_UNDERSCORES)
        return gsonBuilder.create()
    }

    @Provides
    @Singleton
    fun provideChampionDao(context: Context): ChampionDao {
        return AppDatabase.getDatabase(context).championDao()
    }
}
